### ISSUE TITLE
AGP 7.4 is not friendly with Gradle 7.6/8.x wrt usesService

### ISSUE THREAD

**Issue by TWiStErRob** [gradle/android-cache-fix-gradle-plugin#466]
I have these enabled
```kotlin
enableFeaturePreview("STABLE_CONFIGURATION_CACHE")
org.gradle.unsafe.configuration-cache=true
org.gradle.unsafe.configuration-cache-problems=fail
org.gradle.unsafe.configuration-cache.quiet=true
org.gradle.warning.mode=fail
org.gradle.configureondemand=true
org.gradle.parallel=true
```
and I run `gradlew :foo:clean :foo:build --no-build-cache` (to simulate full build) on any library subproject, I see a lot of warnings like this (like a lot lot):
```
> Task :annotations:lintDebug
Build service 'com.android.build.gradle.internal.profile.AnalyticsService_20d9b13a-99f4-4cf4-9e53-419fbc9d243b'
is being used by task ':annotations:lintDebug' without the corresponding declaration via Task#usesService'.
This will fail with an error in Gradle 8.0.
Declare the association between the task and the build service using 'Task#usesService'.
Consult the upgrading guide for further information:
https://docs.gradle.org/7.6/userguide/upgrading_version_7.html#undeclared_build_service_usage
```

This probably also applies to AGP [4.2.x, 7.4.x] range based on the existence of the main violator:
![image](https://user-images.githubusercontent.com/2906988/222835918-fbbb8ba8-4085-4430-be02-3b9ed569921d.png)

Now, I understand this could be a configuration cache and not build cache issue. I'm wondering if it would make sense to develop this idea further: _the build services can be retrofitted onto tasks, making AGP fully compatible with Gradle 7.6 and 8.0._ I'm reporting here, because I think this plugin could host a workaround to this.

<details><summary>Here's a proof of concept (click to expand)</summary>

```kotlin
inline fun <reified T : BuildService<*>> useBuildService(taskName: String) {
	afterEvaluate {
		tasks.named(taskName).configure {
			val buildServiceProvider =
				project
					.gradle
					.sharedServices
					.registrations
					.single { it.name.startsWith(T::class.java.name) }
					.service
			@Suppress("UnstableApiUsage")
			usesService(buildServiceProvider)
		}
	}
}

// This list is generated by running a full build on a library project, and then transforming the warnings to code.
// Probably the same needs to be repeated on an app project too.
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("mergeDebugJniLibFolders")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("copyDebugJniLibsProjectAndLocalJars")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("generateDebugResValues")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("packageDebugResources")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("javaPreCompileDebug")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("mergeDebugShaders")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("packageDebugAssets")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("writeDebugAarMetadata")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("extractDeepLinksDebug")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("extractProguardFiles")
useBuildService<com.android.build.gradle.internal.SdkComponentsBuildService>("parseDebugLocalResources")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("parseDebugLocalResources")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("writeDebugLintModelMetadata")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("processDebugManifest")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("mergeDebugJavaResource")
useBuildService<com.android.build.gradle.internal.SdkComponentsBuildService>("generateDebugRFile")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("generateDebugRFile")
useBuildService<com.android.build.gradle.internal.services.SymbolTableBuildService>("generateDebugRFile")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("extractDebugAnnotations")
useBuildService<com.android.build.gradle.internal.services.LintClassLoaderBuildService>("extractDebugAnnotations")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("compileDebugJavaWithJavac")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("syncDebugLibJars")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("bundleLibRuntimeToJarDebug")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("createFullJarDebug")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("lintAnalyzeDebug")
useBuildService<com.android.build.gradle.internal.services.LintClassLoaderBuildService>("lintAnalyzeDebug")
useBuildService<com.android.build.gradle.internal.ide.dependencies.LibraryDependencyCacheBuildService>("lintAnalyzeDebug")
useBuildService<com.android.build.gradle.internal.ide.dependencies.MavenCoordinatesCacheBuildService>("lintAnalyzeDebug")
useBuildService<com.android.build.gradle.internal.services.StringCachingBuildService>("lintAnalyzeDebug")
useBuildService<com.android.build.gradle.internal.SdkComponentsBuildService>("lintAnalyzeDebug")
useBuildService<com.android.build.gradle.internal.errors.SyncIssueReporterImpl.GlobalSyncIssueService>("lintAnalyzeDebug")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("lintReportDebug")
useBuildService<com.android.build.gradle.internal.services.LintClassLoaderBuildService>("lintReportDebug")
useBuildService<com.android.build.gradle.internal.ide.dependencies.LibraryDependencyCacheBuildService>("lintReportDebug")
useBuildService<com.android.build.gradle.internal.ide.dependencies.MavenCoordinatesCacheBuildService>("lintReportDebug")
useBuildService<com.android.build.gradle.internal.SdkComponentsBuildService>("lintReportDebug")
useBuildService<com.android.build.gradle.internal.errors.SyncIssueReporterImpl.GlobalSyncIssueService>("lintReportDebug")
useBuildService<com.android.build.gradle.internal.profile.AnalyticsService>("lintDebug")
```

This is the first iteration, strangely it doesn't work on first builds, but if that's solved, this can be productionized to support all variants and flavors and stuff, by adding a `println("useBuildService<${this::class.java.name.removeSuffix("_Decorated")}, ${T::class.java.name}>()")` into the `configure { }` block to generate more dynamic code, where:
```
fun <Task, Service> useBuildService() {
    ... similar code as above but with withType.configureEach
```

</details>

**Comment by cdsap** [comment]
Hi @TWiStErRob, first, thank you for the complete description and suggested solution, I really like the Kotlin poc. It would be great to address this issue in the plugin.

Anyways, I'm trying to reproduce the issue in the project `nowinadroid` using Gradle 7.6 and AGP 7.4 with the configuration cache setup you posted, but I'm not seeing the warnings(simulating a full build without cache), I'm not sure what I'm missing, can you provide a simple reproducible example please?

**Comment by runningcode** [comment]
I don't see that this is in the scope of the Android Cache fix plugin. The warning doesn't affect build performance. Thanks for bringing this to our attention.

**Comment by TWiStErRob** [comment]
Ah, I see, similar warnings were actually affecting the build and hence were fixed in this plugin, but this is just spamming warnings on "unsupported" AGP/Gradle combos.

**Comment by TWiStErRob** [comment]
This came up again. Same thing on AGP 8.1.1 on Gradle 8.2.1 and even on AGP 8.2.0-beta04 on Gradle 8.2.1. So it's not just incompatible Gradle versions, AGP simply doesn't register its services with `usesService` which means it's really hard to see if other plugins are using it or not.

https://github.com/gradle/android-cache-fix-gradle-plugin/issues/466#issuecomment-1544044152 is still valid, it doesn't affect this plugin, raised https://issuetracker.google.com/issues/300617088.

**Comment by runningcode** [comment]
Thanks for raising this with Google!
